Kernel/Graphics/VirtIOGPU/VirtIOGPU.cpp

```diff
@@ -0,0 +1,286 @@
+    VERIFY(queue_index == CONTROLQ);
+
+    auto& queue = get_queue(CONTROLQ);
+    ScopedSpinLock queue_lock(queue.lock());
+    queue.discard_used_buffers();
+    m_has_outstanding_request = false;
+}
+
+u32 VirtIOGPU::get_pending_events()
+{
+    return config_read32(*m_device_configuration, DEVICE_EVENTS_READ);
+}
+
+void VirtIOGPU::clear_pending_events(u32 event_bitmask)
+{
+    config_write32(*m_device_configuration, DEVICE_EVENTS_READ, event_bitmask);
+}
+
+void VirtIOGPU::query_display_information()
+{
+    Locker locker(m_operation_lock);
+    auto& request = *reinterpret_cast<VirtIOGPUCtrlHeader*>(m_scratch_space->vaddr().as_ptr());
+    auto& response = *reinterpret_cast<VirtIOGPURespDisplayInfo*>((m_scratch_space->vaddr().offset(sizeof(request)).as_ptr()));
+
+    populate_virtio_gpu_request_header(request, VirtIOGPUCtrlType::VIRTIO_GPU_CMD_GET_DISPLAY_INFO, VIRTIO_GPU_FLAG_FENCE);
+
+    synchronous_virtio_gpu_command(start_of_scratch_space(), sizeof(request), sizeof(response));
+
+    for (size_t i = 0; i < VIRTIO_GPU_MAX_SCANOUTS; ++i) {
+        auto& scanout = response.scanout_modes[i];
+        if (!scanout.enabled)
+            continue;
+        dbgln("Scanout {}: x: {}, y: {}, width: {}, height: {}", i, scanout.rect.x, scanout.rect.y, scanout.rect.width, scanout.rect.height);
+        m_display_info = scanout;
+        m_chosen_scanout = i;
+    }
+    m_display_info.rect.width = 1024;
+    m_display_info.rect.height = 768;
+}
+
+void VirtIOGPU::create_framebuffer()
+{
+    Locker locker(m_operation_lock);
+    auto& request = *reinterpret_cast<VirtIOGPUResourceCreate2D*>(m_scratch_space->vaddr().as_ptr());
+    auto& response = *reinterpret_cast<VirtIOGPUCtrlHeader*>((m_scratch_space->vaddr().offset(sizeof(request)).as_ptr()));
+
+    populate_virtio_gpu_request_header(request.header, VirtIOGPUCtrlType::VIRTIO_GPU_CMD_RESOURCE_CREATE_2D, VIRTIO_GPU_FLAG_FENCE);
+
+    request.resource_id = m_framebuffer_id = 1;
+    request.width = m_display_info.rect.width;
+    request.height = m_display_info.rect.height;
+    request.format = static_cast<u32>(VirtIOGPUFormats::VIRTIO_GPU_FORMAT_B8G8R8X8_UNORM);
+
+    synchronous_virtio_gpu_command(start_of_scratch_space(), sizeof(request), sizeof(response));
+
+    VERIFY(response.type == static_cast<u32>(VirtIOGPUCtrlType::VIRTIO_GPU_RESP_OK_NODATA));
+    dbgln_if(VIRTIO_DEBUG, "VirtIOGPU: Allocated resource id");
+}
+
+void VirtIOGPU::attach_backing_storage()
+{
+    Locker locker(m_operation_lock);
+    // Allocate backing region
+    size_t buffer_length = framebuffer_size_in_bytes();
+    m_framebuffer = MM.allocate_kernel_region(page_round_up(buffer_length), "VirtGPU FrameBuffer", Region::Access::Read | Region::Access::Write, AllocationStrategy::AllocateNow);
+    VERIFY(!!m_framebuffer);
+    size_t num_mem_regions = page_round_up(buffer_length) / PAGE_SIZE;
+    VERIFY(num_mem_regions == m_framebuffer->vmobject().page_count());
```

Comment:
this VERIFY seems useless, when can it fail?

Reply:
Ah, this VERIFY actually is useless, I added it while debugging to make sure I got my calculations right. Will remove in next revision.